lib/src/third_party/diff_match_patch/utils.dart

```diff
@@ -0,0 +1,253 @@
+  final n = min(text1.length, text2.length);
+
+  for (var i = 0; i < n; i++) {
+    if (text1[i] != text2[i]) {
+      return i;
+    }
+  }
+  return n;
+}
+
+/// Determine the common suffix of two strings.
+///
+/// [text1] is the first string. [text2] is the second string.
+/// Returns the number of characters common to the end of each string.
+@visibleForTesting
+int diffCommonSuffix(String text1, String text2) {
+  // TODO: Once Dart's performance stabilizes, determine if linear or binary
+  // search is better.
+  // Performance analysis: https://neil.fraser.name/news/2007/10/09/
+  final text1Length = text1.length;
+  final text2Length = text2.length;
+  final n = min(text1Length, text2Length);
+
+  for (var i = 1; i <= n; i++) {
+    if (text1[text1Length - i] != text2[text2Length - i]) {
+      return i - 1;
+    }
+  }
+  return n;
+}
+
+/// Split a text into a list of strings. Reduce the texts to a string of
+/// hashes where each Unicode character represents one line.
+///
+/// [text] is the string to encode.
+/// [lineArray] is a List of unique strings.
+/// [lineHash] is a Map of strings to indices.
+/// [maxLines] is the maximum length for lineArray.
+/// Returns an encoded string.
+String _diffLinesToCharsMunge(
+  String text,
+  List<String> lineArray,
+  Map<String, int> lineHash,
+  int maxLines,
+) {
+  var lineStart = 0;
+  var lineEnd = -1;
+  String line;
+  final chars = StringBuffer();
+
+  // Walk the text, pulling out a substring for each line.
+  // text.split('\n') would would temporarily double our memory footprint.
+  // Modifying text would create many large strings to garbage collect.
+  while (lineEnd < text.length - 1) {
+    lineEnd = text.indexOf('\n', lineStart);
+
+    if (lineEnd == -1) {
+      lineEnd = text.length - 1;
+    }
+    line = text.substring(lineStart, lineEnd + 1);
+
+    if (lineHash.containsKey(line)) {
+      chars.writeCharCode(lineHash[line]!);
+    } else {
+      if (lineArray.length == maxLines) {
+        // Bail out at 65535 because
+        // final chars1 =  StringBuffer();
+        // chars1.writeCharCode(65536);
+        // chars1.toString().codeUnitAt(0) == 55296;
+        line = text.substring(lineStart);
+        lineEnd = text.length;
+      }
+      lineArray.add(line);
+      lineHash[line] = lineArray.length - 1;
+      chars.writeCharCode(lineArray.length - 1);
+    }
+    lineStart = lineEnd + 1;
+  }
+  return chars.toString();
+}
+
+/// Split two texts into a list of strings. Reduce the texts to a string of
+/// hashes where each Unicode character represents one line.
+///
+/// [text1] is the first string.
+/// [text2] is the second string.
+/// Returns a Map containing the encoded text1, the encoded text2 and
+/// the List of unique strings. The zeroth element of the List of
+/// unique strings is intentionally blank.
+@visibleForTesting
+Map<String, dynamic> diffLinesToChars(String text1, String text2) {
```

Comment:
I don't think this is called anywhere.

Reply:
Yes, but maybe we can use them as debugging tools, there are some other functions like this in diffs that are not used anywhere, should I remove all of them?

Reply:
I would think so unless they have a concrete purpose. It clutters the review with untested code. You can always reintroduce them if they are needed.

Reply:
Okay, I'll remove all the unused functions.